File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int k = sc.nextInt();

        int max = 200000;
        int[] minTime = new int[max + 1];
        Arrays.fill(minTime, -1);

        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.add(n);
        minTime[n] = 0;

        while (!q.isEmpty()) {
            int cur = q.poll();

            if (cur == k) {
                System.out.println(minTime[k]);
                return;
            }

            if (cur * 2 <= max && minTime[cur * 2] == -1) {
                minTime[cur * 2] = minTime[cur];
                q.addFirst(cur * 2);
            }

            if (cur - 1 >= 0 && minTime[cur - 1] == -1) {
                minTime[cur - 1] = minTime[cur] + 1;
                q.addLast(cur - 1);
            }

            if (cur + 1 <= max && minTime[cur + 1] == -1) {
                minTime[cur + 1] = minTime[cur] + 1;
                q.addLast(cur + 1);
            }
        }
    }
}

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.kt
package com.example.myapplication
fun main(){
    var(a,b) = readln().split(" ").map { it.toInt() }
    var visitied = BooleanArray(200001){false}
    var pos = IntArray(200001){Integer.MAX_VALUE}
    var q = ArrayDeque<Int>()
    q.add(a)
    visitied[a] = true
    var time = 0
    loop@while(!q.isEmpty()){
        var s = q.size
        while(s-- > 0){
            val cur = q.removeFirst()
            if(cur == b)
                break@loop
            if(cur > b){ // 도착 위치 넘어갔으면
                if(!visitied[cur-1]){
                    visitied[cur-1] = true;
                    q.add(cur-1);
                }
                continue;
            }else{
                if(cur*2 < 200000 && !visitied[cur*2]){
                    visitied[cur*2] = true;
                    q.addFirst(cur*2);
                    s++
                }
                if(cur-1 > 0 && !visitied[cur-1]){
                    visitied[cur-1] = true;
                    q.add(cur-1);
                }
                if(!visitied[cur+1]){
                    visitied[cur+1] = true;
                    q.add(cur+1);
                }
            }
        }
        time++
    }
    println(time)
}

